### PR TITLE
Horizontally, Vertically & All with margin parameter

### PR DESCRIPTION
**Pull request by OlivierPineau**

### CODE REVIEW

File: Sources/Impl/PinLayoutImpl+Coordinates.swift [hunk 1]
```diff
@@ -24,6 +24,19 @@ extension PinLayoutImpl {
     internal func top(_ context: Context) {
         setTop(0, context)
     }
+
+    @discardableResult
+    internal func top(_ value: CGFloat, _ context: Context) -> PinLayout {
+        setTop(value, context)
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl+Coordinates.swift]: You don't need to add this method if you call `setTop(value, context), setBottom(value, context) ...` in `func all(_ value: CGFloat)`

Comment (lucdion) [reply Sources/Impl/PinLayoutImpl+Coordinates.swift]: Ah, ok you can keep it that way, I use this pattern somewhere else 😊

Comment (OlivierPineau) [reply Sources/Impl/PinLayoutImpl+Coordinates.swift]: :)

Comment (lucdion) [reply Sources/Impl/PinLayoutImpl+Coordinates.swift]: But could you update the method `func top(_ value: CGFloat) -> PinLayout` in PinLayoutImpl.swift to:
```
@discardableResult
    func top(_ value: CGFloat) -> PinLayout {
        return top(value, { return "top(\(value))" })
    }
```

File: Sources/Impl/PinLayoutImpl+Coordinates.swift [hunk 2]
```diff
@@ -24,6 +24,19 @@ extension PinLayoutImpl {
...
+        return self
+    }
+
+    @discardableResult
+    internal func top(_ percent: Percent, _ context: Context) -> PinLayout {
+        guard let layoutSuperviewRect = layoutSuperviewRect(context) else { return self }
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl+Coordinates.swift]: Could you update the method `func top(_ percent: Percent)-> PinLayout` in PinLayoutImpl.swift to:
```
func top(_ percent: Percent) -> PinLayout {
        return top(percent,  { return "top(\(percent.description))" })
    }
```

File: Sources/Impl/PinLayoutImpl+Coordinates.swift [hunk 3]
```diff
@@ -134,6 +147,13 @@ extension PinLayoutImpl {
         setBottom(layoutSuperviewRect.height - value, context)
         return self
     }
+
+    @discardableResult
+    internal func bottom(_ percent: Percent, _ context: Context) -> PinLayout {
+        guard let layoutSuperviewRect = layoutSuperviewRect(context) else { return self }
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl+Coordinates.swift]: Could you update the method `func bottom(_ percent: Percent) -> PinLayout` in PinLayoutImpl.swift to:
```
func bottom(_ percent: Percent) -> PinLayout {
        return bottom(percent, { return "bottom(\(percent.description))" })
    }
```

File: Sources/Impl/PinLayoutImpl.swift [hunk 4]
```diff
@@ -234,20 +234,63 @@ class PinLayoutImpl: PinLayout {
         left({ "all() left coordinate" })
         return self
     }
+
+    @discardableResult
+    func all(_ value: CGFloat) -> PinLayout {
+        func context() -> String { return "all(\(value))" }
+        top(value, context)
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl.swift]: Could you do as in `all()`, i.e. set a different context string for all internal calls (ex: { "all(\(value)) top coordinate" })

File: Sources/Impl/PinLayoutImpl.swift [hunk 5]
```diff
@@ -234,20 +234,63 @@ class PinLayoutImpl: PinLayout {
...
     func horizontally() -> PinLayout {
         right({ "horizontally() right coordinate" })
         left({ "horizontally() left coordinate" })
         return self
     }
+
+    @discardableResult
+    func horizontally(_ value: CGFloat) -> PinLayout {
+        func context() -> String { return "horizontally(\(value))" }
+        left(value, context)
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl.swift]: Could you do as in `all()`, i.e. set a different context string for all internal calls (ex: `{ "horizontally(\(value)) left coordinate" }`)

File: Sources/Impl/PinLayoutImpl.swift [hunk 6]
```diff
@@ -234,20 +234,63 @@ class PinLayoutImpl: PinLayout {
...
+        right(value, context)
+        return self
+    }
+
+    @discardableResult
+    func horizontally(_ percent: Percent) -> PinLayout {
+        func context() -> String { return "horizontally(\(percent.description))" }
+        left(percent, context)
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl.swift]: Could you do as in `all()`, i.e. set a different context string for all internal calls (ex: `{ "horizontally(\(value)) left coordinate" }`)

File: Sources/Impl/PinLayoutImpl.swift [hunk 7]
```diff
@@ -234,20 +234,63 @@ class PinLayoutImpl: PinLayout {
...
     func vertically() -> PinLayout {
         top({ "vertically() top coordinate" })
         bottom({ "vertically() bottom coordinate" })
         return self
     }
 
+    @discardableResult
+    func vertically(_ value: CGFloat) -> PinLayout {
+        func context() -> String { return "vertically(\(value))" }
+        top(value, context)
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl.swift]: Could you do as in `all()`, i.e. set a different context string for all internal calls (ex: `{ "vertically(\(value)) top coordinate" }`)

File: Sources/Impl/PinLayoutImpl.swift [hunk 8]
```diff
@@ -234,20 +234,63 @@ class PinLayoutImpl: PinLayout {
...
+        bottom(value, context)
+        return self
+    }
+
+    @discardableResult
+    func vertically(_ percent: Percent) -> PinLayout {
+        func context() -> String { return "vertically(\(percent.description))" }
+        top(percent, context)
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl.swift]: Could you do as in `all()`, i.e. set a different context string for all internal calls (ex: `{ "vertically(\(value)) top coordinate" }`)

File: Tests/PinEdgesSpec.swift [hunk 9]
```diff
@@ -531,6 +561,22 @@ class PinEdgesSpec: QuickSpec {
                 aView.pin.right(20).horizontally()
                 expect(Pin.lastWarningText).to(contain(["horizontally() right coordinate", "won't be applied", "already been set to 20"]))
             }
+            it("should warn") {
+                aView.pin.left(10).horizontally(20)
+                expect(Pin.lastWarningText).to(contain(["horizontally(20.0)", "won't be applied", "already been set to 10"]))
```
Comment (lucdion) [inline Tests/PinEdgesSpec.swift]: Once you have updated the method context, thanks to updating the expectation, here it should contain `left`.

File: Tests/PinEdgesSpec.swift [hunk 10]
```diff
@@ -565,6 +621,22 @@ class PinEdgesSpec: QuickSpec {
                 aView.pin.bottom(20).vertically()
                 expect(Pin.lastWarningText).to(contain(["vertically() bottom coordinate", "won't be applied", "already been set to 20"]))
             }
+            it("should warn") {
+                aView.pin.top(10).vertically(20)
```
Comment (lucdion) [inline Tests/PinEdgesSpec.swift]: Once you have updated the method context, thanks to updating expectations, here it should contain `top`.

File: Sources/Impl/PinLayoutImpl.swift [hunk 11]
```diff
@@ -234,20 +230,58 @@ class PinLayoutImpl: PinLayout {
...
     @discardableResult
     func vertically() -> PinLayout {
         top({ "vertically() top coordinate" })
         bottom({ "vertically() bottom coordinate" })
         return self
     }
 
+    @discardableResult
+    func vertically(_ value: CGFloat) -> PinLayout {
+        top(value, { return "vertically(\(value)) top" })
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl.swift]: Just  missing the "coordinate" word: "vertically(\(value)) top coordinate".
Also, for the bottom.

File: Sources/Impl/PinLayoutImpl.swift [hunk 12]
```diff
@@ -234,20 +230,58 @@ class PinLayoutImpl: PinLayout {
...
+        bottom(value, { return "vertically(\(value)) bottom" })
+        return self
+    }
+
+    @discardableResult
+    func vertically(_ percent: Percent) -> PinLayout {
+        top(percent, { return "vertically(\(percent.description)) top" })
```
Comment (lucdion) [inline Sources/Impl/PinLayoutImpl.swift]: Just  missing the "coordinate" word: "vertically(\(value)) top coordinate".
Also, for the bottom.

**Pull request merged**